va/caa_test.go

```diff
@@ -578,6 +640,322 @@ func TestIsCAAValidParams(t *testing.T) {
+			domains:           "satisfiable-wildcard.com",
+			maxLookupFailures: 1,
+			expectedProb:      probs.CAA("During secondary CAA rechecking: While processing CAA for satisfiable-wildcard.com: CAA record for satisfiable-wildcard.com prevents issuance"),
+			expectedDiffLog:   `INFO: remoteVADifferentials JSON={"Domain":"satisfiable-wildcard.com","AccountID":1,"ChallengeType":"dns-01","PrimaryResult":null,"RemoteSuccesses":0,"RemoteFailures":[{"VAHostname":"hijacked","Problem":{"type":"caa","detail":"While processing CAA for satisfiable-wildcard.com: CAA record for satisfiable-wildcard.com prevents issuance"}},{"VAHostname":"hijacked","Problem":{"type":"caa","detail":"While processing CAA for satisfiable-wildcard.com: CAA record for satisfiable-wildcard.com prevents issuance"}},{"VAHostname":"hijacked","Problem":{"type":"caa","detail":"While processing CAA for satisfiable-wildcard.com: CAA record for satisfiable-wildcard.com prevents issuance"}}]}`,
+			remoteVAs: []RemoteVA{
+				{hijackedVA1, hijackedUA},
+				{hijackedVA2, hijackedUA},
+				{hijackedVA3, hijackedUA},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			va, mockLog := setup(nil, tc.maxLookupFailures, localUA, tc.remoteVAs)
+			defer mockLog.Clear()
+
+			if tc.localVADNSClientOverride == nil {
+				va.dnsClient = caaMockDNS{}
+			} else {
+				va.dnsClient = *tc.localVADNSClientOverride
+			}
+
+			// MultiVAFullResults: false is inherently flaky because of the
+			// non-deterministic nature of concurrent goroutine returns. We,
+			// boulder dev, made a decision to skip testing that path and
+			// eventually make MultiVAFullResults: true the default.
+			features.Set(features.Config{
+				RVACAARechecking:   true,
+				EnforceMultiVA:     true,
+				MultiVAFullResults: true,
+			})
+			defer features.Reset()
+
+			isValidRes, err := va.IsCAAValid(context.TODO(), &vapb.IsCAAValidRequest{
+				Domain:           tc.domains,
+				ValidationMethod: string(core.ChallengeTypeDNS01),
+				AccountURIID:     1,
+			})
+			test.AssertNotError(t, err, "Should not have errored, but did")
+
+			if tc.expectedProb != nil {
+				test.AssertEquals(t, isValidRes.Problem.Detail, tc.expectedProb.Detail)
+			} else if isValidRes.Problem != nil {
+				test.AssertNotNil(t, tc.expectedProb, "Test should have an expectedProb")
+			}
+
+			var invalidRVACount int
+			for _, x := range va.remoteVAs {
+				if x.Address == "broken" || x.Address == "hijacked" {
+					invalidRVACount++
+				}
+			}
+
+			gotRequestProbs := mockLog.GetAllMatching(".IsCAAValidRequest returned problem: ")
+			test.AssertEquals(t, len(gotRequestProbs), invalidRVACount)
+
+			gotDifferential := mockLog.GetAllMatching("remoteVADifferentials JSON=.*")
+			if features.Get().MultiVAFullResults && tc.expectedDiffLog != "" {
+				test.AssertEquals(t, len(gotDifferential), 1)
+				test.AssertEquals(t, gotDifferential[0], tc.expectedDiffLog)
```

Comment:
Similarly, I don't love asserting that the log line is exactly what we expect. It makes for very unwieldy and fragile test expectations. I'm not sure what the best alternative here is, since the log line carries a lot of information, so maybe it's fine in this case, but I'd like to think about alternatives.